codecs/avif/enc/avif_enc.cpp

```diff
@@ -3,15 +3,31 @@
 #include <emscripten/val.h>
 #include "avif/avif.h"
 
+#include <string>
+
+#define RETURN_NULL_IF_NOT_EQUALS(val1, val2) \
+  do {                                        \
+    if (val1 != val2)                         \
+      return val::null();                     \
+  } while (false)
+#define RETURN_NULL_IF_EQUALS(val1, val2) \
+  do {                                    \
+    if (val1 == val2)                     \
+      return val::null();                 \
+  } while (false)
+
 using namespace emscripten;
 
+using AvifImagePtr = std::unique_ptr<avifImage, decltype(&avifImageDestroy)>;
+using AvifEncoderPtr = std::unique_ptr<avifEncoder, decltype(&avifEncoderDestroy)>;
+
 struct AvifOptions {
-  // [0 - 63]
-  // 0 = lossless
-  // 63 = worst quality
-  int cqLevel;
-  // As above, but -1 means 'use cqLevel'
-  int cqAlphaLevel;
+  // [0 - 100]
+  // 0 = worst quality
+  // 100 = lossless
+  int quality;
+  // As above, but -1 means 'use quality'
+  int qualityAlpha;
```

Comment:
In a follow-up pull request, we should look into using the new `autoTiling` option of the AVIF encoder.

I suggest Squoosh set `autoTiling` to `AVIF_TRUE` by default. (The default of `autoTiling` in libavif is `AVIF_FALSE` for backward compatibility.) The Squoosh UI can have a "manual tiling" box, which when checked,  shows the `tileRowsLog2` and `tileColsLog2` fields (both default to 0, meaning no tiling).

Reply:
Gotcha, I'll try to add it to this PR or at least make raise an issue for it. I assume `autoTiling` automatically chooses `tileRowsLog2` and `tileColsLog2`, is there any more documentation about it that I can link to?

Reply:
Unless this requires just simple changes, I suggest filing an issue for this and implementing this in a follow-up pull request. The review of this pull request has been going on for a while. We need to watch out for "code review fatigue."

You understand `autoTiling` correctly. The `autoTiling` algorithm may be tweaked in the future, so we didn't document the current algorithm in the public header avif/avif.h. You can find the current algorithm in libavif/src/write.c:

```
    if (encoder->autoTiling) {
        // Use as many tiles as allowed by the minimum tile area requirement and impose a maximum
        // of 8 tiles.
        const int threads = 8;
        avifSetTileConfiguration(threads, firstCell->width, firstCell->height, &encoder->data->tileRowsLog2, &encoder->data->tileColsLog2);
    }
```

I do think all the automatic tiling algorithms have some common elements, such as imposing a minimum tile area or tile width and height, and potentially imposing a maximum number of tiles. But I am not sure if you need to mention this. You can, however, say that the use of tiles makes it easier for the decoder to decode an AVIF image with multiple threads in parallel.